File: shells/android/app/src/main/java/com/feathersurf/browser/FeatherSurfNative.kt
package com.feathersurf.browser

/**
 * JNI bridge to the FeatherSurf Rust core.
 *
 * The native library (libfeathersurf_ffi.so) is loaded at class init time.
 * All methods are thin wrappers around the C FFI functions defined in ffi/src/lib.rs.
 */
object FeatherSurfNative {

    init {
        System.loadLibrary("feathersurf_ffi")
    }

    // -- Native methods (implemented in Rust via JNI) --

    /**
     * Initialize the Rust core. Must be called once before any other method.
     */
    external fun nativeInit()

    /**
     * Create a new tab with the given URL.
     * Returns the tab ID (u64).
     */
    external fun nativeTabCreate(url: String): Long

    /**
     * Destroy a tab and free its resources.
     */
    external fun nativeTabDestroy(tabId: Long)

    /**
     * Get the current state of a tab.
     * Returns a TabState constant.
     */
    external fun nativeTabState(tabId: Long): Int

    /**
     * Transition a tab to a new state.
     * Returns 0 on success, non-zero on failure.
     */
    external fun nativeTabTransition(tabId: Long, targetState: Int): Int

    /**
     * Get the current URL of a tab.
     */
    external fun nativeTabUrl(tabId: Long): String

    /**
     * Get the current title of a tab.
     */
    external fun nativeTabTitle(tabId: Long): String

    // -- Kotlin-friendly wrappers --

    fun init() {
        nativeInit()
    }

    fun tabCreate(url: String): Long {
        return nativeTabCreate(url)
    }

    fun tabDestroy(tabId: Long) {
        nativeTabDestroy(tabId)
    }

    fun tabState(tabId: Long): Int {
        return nativeTabState(tabId)
    }

    fun tabTransition(tabId: Long, targetState: Int): Boolean {
        return nativeTabTransition(tabId, targetState) == 0
    }

    fun tabUrl(tabId: Long): String {
        return nativeTabUrl(tabId)
    }

    fun tabTitle(tabId: Long): String {
        return nativeTabTitle(tabId)
    }
}

File: shells/ios/FeatherSurf/FeatherSurfNative.swift
import Foundation

/// C FFI bridge to the FeatherSurf Rust core.
///
/// The static library (libfeathersurf_ffi.a) is linked at build time.
/// This module provides Swift-friendly wrappers around the C functions.
class FeatherSurfNative {

    // MARK: - C Function Declarations

    /// Opaque handle to a Tab instance (from Rust).
    typealias FfiTab = OpaquePointer

    /// Create a new tab.
    /// - Parameter url: The initial URL string.
    /// - Returns: An opaque handle to the tab, or nil on failure.
    static func tabCreate(url: String) -> FfiTab? {
        return url.withCString { cUrl in
            feathersurf_tab_create(0, cUrl)
        }
    }

    /// Destroy a tab and free its memory.
    /// - Parameter tab: The tab handle.
    static func tabDestroy(_ tab: FfiTab) {
        feathersurf_tab_destroy(tab)
    }

    /// Get the current state of a tab.
    /// - Parameter tab: The tab handle.
    /// - Returns: A TabState value.
    static func tabState(_ tab: FfiTab) -> TabState {
        let raw = feathersurf_tab_state(tab)
        return TabState(rawValue: raw) ?? .active
    }

    /// Transition a tab to a new state.
    /// - Parameters:
    ///   - tab: The tab handle.
    ///   - state: The target state.
    /// - Returns: true on success, false on failure.
    static func tabTransition(_ tab: FfiTab, to state: TabState) -> Bool {
        return feathersurf_tab_transition(tab, state.rawValue) == 0
    }

    /// Get the current URL of a tab.
    /// - Parameter tab: The tab handle.
    /// - Returns: The URL string, or nil.
    static func tabUrl(_ tab: FfiTab) -> String? {
        guard let cStr = feathersurf_tab_url(tab) else { return nil }
        return String(cString: cStr)
    }

    /// Get the current title of a tab.
    /// - Parameter tab: The tab handle.
    /// - Returns: The title string, or nil.
    static func tabTitle(_ tab: FfiTab) -> String? {
        guard let cStr = feathersurf_tab_title(tab) else { return nil }
        let result = String(cString: cStr)
        feathersurf_string_free(cStr)
        return result
    }
}
